File: i3_nix_playerctl/playerctl.py
#!/usr/bin/env python
import os
import sys
import time

MAXLEN = 30

def getTitle(player):
    cmdTitle = "playerctl -p %s metadata title" % player # prepare command
    title = os.popen(cmdTitle)                           # execute it
    return title.read().strip()                          # read the output and strip it 

def getStatus(player):
    cmdStatus = "playerctl -p %s status" % player        # prepaer command
    status = os.popen(cmdStatus)                         # execute it
    return status.read().strip()                         # read the output and strip it 

def getVolume(player):
    cmdVolume = "playerctl volume -p %s" % player        # prepaer command
    volume = os.popen(cmdVolume)                         # execute it
    volume = volume.read().strip()                       # read the output and strip it 
    volume = int(float(volume) * 100)                    # we don't want decimal number
    return str(volume) # we can't concatenate string with float, so return string

def printSongTitle(title, volume=None):
    maxLen = MAXLEN - 3 # for player chars

    l = len(title)
    if volume is not None:
        l = l + 3
    if l == 0:
        print(" ")
    elif l > maxLen:
        i = int(time.time()%(l-maxLen))
        if i + maxLen >= l:
            i = l - maxLen 
        if volume is not None:
            print(title[i:i+maxLen+1-3] + " " +volume)
        else:
            print(title[i:i+maxLen+1])
    else:
        if volume is not None:
            print(title)
        else:
            print(title + " " + volume)
    

def printColor(status):
    if status == "Playing":
        print("#00AFFF")
        print("#00AFFF")
    elif status == "Paused":
        print("#9616DB")
        print("#9616DB")
    else:
        print("#FF0000")
        print("#FF0000")

def defaultPrint(title, player):
    printSongTitle(title)
    # wait for player to register status 
    # change so it doesn't look laggy in bar
    # or else it will wait next second to
    # change color
    time.sleep(0.05) 
    print(getStatus(player))
    printColor(getStatus(player))
    

try:
    player = os.environ["BLOCK_INSTANCE"]
except:
    player = "audacious"
try:
    buttonValue = os.environ["BLOCK_BUTTON"]
except:
    buttonValue = ""

status = getStatus(player)
title = getTitle(player)

if player == "audacious" and (title is None or title == "(null)" or title == ""):
    title = ""
    if buttonValue != '':
        if buttonValue == "1":
            o = os.popen("audacious")


if buttonValue != '':
    if buttonValue == "1":
        os.popen("playerctl -p %s play-pause" %player)
        defaultPrint(title, player)
    elif buttonValue == "2" and player == "audacious":
        os.popen("killall audacious")
        defaultPrint(title, player)
    elif buttonValue == "3":
        os.popen("playerctl -p %s stop" %player)
        defaultPrint(title, player)
    elif buttonValue == "4":
        os.popen("playerctl -p %s volume 0.05+" %player) # if step is higher, it won't go near 100%
        printSongTitle(title, getVolume(player))
        printColor(status)
    elif buttonValue == "5":
        os.popen("playerctl -p %s volume 0.05-" %player) # if step is higher, it won't go near 100%
        printSongTitle(title, getVolume(player))
        printColor(status)
    elif buttonValue == "8":
        os.popen("playerctl -p %s next" %player)
        defaultPrint(title, player)
    elif buttonValue == "9":
        os.popen("playerctl -p %s previous" %player)
        defaultPrint(title, player)
else:
    defaultPrint(title, player)

File: i3_nix_playerctl/playerctl.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
	"time"
)

const MAXLEN = 30

func executeCommand(cmd string) string {
	var command string
	args := strings.Split(cmd, " ") // first arg is acctual command
	if len(args) > 1 {
		command = args[0]
		args = args[1:]
	} else {
		command = cmd
	}
	out, err := exec.Command(command, args...).Output()
	if err != nil {
		log.Fatal(err)
	}
	return strings.Trim(string(out), " \n")
}

func getTitle(player string) string {
	cmd := fmt.Sprintf("playerctl -p %s metadata title", player)
	out := executeCommand(cmd)
	if out == "(null)" {
		out = ""
	}
	return out
}

func getStatus(player string) string {
	cmd := fmt.Sprintf("playerctl -p %s status", player)
	out := executeCommand(cmd)
	return out
}

func getVolume(player string) string {
	args := strings.Split(fmt.Sprintf("volume -p %s", player), " ")
	out, err := exec.Command("playerctl", args...).Output()
	if err != nil {
		log.Fatal(err)
	}
	return strings.Trim(string(out), " \n")
}

func printSongTitle(title, volume string) {
	var maxLen int64
	maxLen = MAXLEN - 3 //for player chars

	l := int64(len(title))
	if volume != "" {
		l = l + 3
	}

	if l == 0 { // if empty
		fmt.Println(" ")
	} else if l > maxLen { // if len is greater than maxlen
		i := time.Now().Unix() % (l - maxLen)

		if i+maxLen >= l {
			i = l - maxLen
		}

		if volume == "" {
			fmt.Println(title[i : i+maxLen+1])
		} else {
			fmt.Println(title[i:i+maxLen+1-3] + " " + volume)
		}
	} else { // if len is ok
		if volume == "" {
			fmt.Println(title)
		} else {
			fmt.Println(title + " " + volume)
		}
	}
}

func printColor(status string) {
	switch status {
	case "Playing":
		fmt.Println("#00AFFF")
		fmt.Println("#00AFFF")
	case "Paused":
		fmt.Println("#9616DB")
		fmt.Println("#9616DB")
	default:
		fmt.Println("#FF0000")
		fmt.Println("#FF0000")
	}
}

func defaultPrint(title, player string) {
	printSongTitle(title, "")
	printColor(getStatus(player))
}

func main() {
	player := os.Getenv("BLOCK_INSTANCE")
	buttonValue := os.Getenv("BLOCK_BUTTON")

	if player == "" {
		player = "audacious"
	}

	status := getStatus(player)
	title := getTitle(player)

	if buttonValue != "" {
		switch buttonValue {
		case "1":
			executeCommand(fmt.Sprintf("playerctl -p %s play-pause", player))
			defaultPrint(title, player)
		case "2":
			if player == "audacious" {
				executeCommand(fmt.Sprintf("killall audacious"))
				defaultPrint(title, player)
			}
		case "3":
			executeCommand(fmt.Sprintf("playerctl -p %s stop", player))
			defaultPrint(title, player)
		case "4":
			executeCommand(fmt.Sprintf("playerctl -p %s volume 0.05+", player)) // if step is higher, it won't go near 100%
			printSongTitle(title, getVolume(player))
			printColor(status)
		case "5":
			executeCommand(fmt.Sprintf("playerctl -p %s volume 0.05-", player)) // if step is higher, it won't go near 100%
			printSongTitle(title, getVolume(player))
			printColor(status)
		case "8":
			executeCommand(fmt.Sprintf("playerctl -p %s next", player))
			defaultPrint(title, player)
		case "9":
			executeCommand(fmt.Sprintf("playerctl -p %s previous", player))
			defaultPrint(title, player)
		}
	} else {
		defaultPrint(title, player)
	}
}
